### DISCUSSION TITLE
Clip and snapshot not available for some events

### DISCUSSION THREAD

**Discussion by fl42** [blakeblackshear/frigate#14617] [Ask A Question]
Hello,

I'm using the `events` MQTT topic to trigger workflows when objects are detected, such as pushing the clip/snapshot to a remote storage.
Frigate is configured to keep all recording segments and snapshots for 30 days.
I'm using 0.14.1-f4f3cfa.

It works nicely for most events, but for some events the clip and snapshot are not available: HTTP API returns 404 when trying to retrieve clip/snapshot for the event, and MQTT event has `has_clip: false` and `has_snapshot: false`.

What could be wrong?

Thank you for your help ;)

Below an example of an event:
```
{"before": {"id": "1729980086.225972-v2uw7d", "camera": "AAA_000023", "frame_time": 1729980086.225972, "snapshot": null, "label": "bird", "sub_label": null, "top_score": 0.0, "false_positive": true, "start_time": 1729980086.225972, "end_time": null, "score": 0.7367673516273499, "box": [811, 503, 839, 516], "area": 364, "ratio": 2.1538461538461537, "region": [678, 400, 998, 720], "stationary": false, "motionless_count": 0, "position_changes": 0, "current_zones": [], "entered_zones": [], "has_clip": false, "has_snapshot": false, "attributes": {}, "current_attributes": []}, "after": {"id": "1729980086.225972-v2uw7d", "camera": "AAA_000023", "frame_time": 1729980086.225972, "snapshot": {"frame_time": 1729980087.585224, "box": [811, 503, 839, 516], "area": 364, "region": [678, 400, 998, 720], "score": 0.7367673516273499, "attributes": []}, "label": "bird", "sub_label": null, "top_score": 0.7367673516273499, "false_positive": false, "start_time": 1729980086.225972, "end_time": null, "score": 0.7367673516273499, "box": [811, 503, 839, 516], "area": 364, "ratio": 2.1538461538461537, "region": [678, 400, 998, 720], "stationary": false, "motionless_count": 0, "position_changes": 0, "current_zones": ["proche"], "entered_zones": ["proche"], "has_clip": false, "has_snapshot": false, "attributes": {}, "current_attributes": []}, "type": "new"}
{"before": {"id": "1729980086.225972-v2uw7d", "camera": "AAA_000023", "frame_time": 1729980086.225972, "snapshot": {"frame_time": 1729980087.585224, "box": [811, 503, 839, 516], "area": 364, "region": [678, 400, 998, 720], "score": 0.7367673516273499, "attributes": []}, "label": "bird", "sub_label": null, "top_score": 0.7367673516273499, "false_positive": false, "start_time": 1729980086.225972, "end_time": null, "score": 0.7367673516273499, "box": [811, 503, 839, 516], "area": 364, "ratio": 2.1538461538461537, "region": [678, 400, 998, 720], "stationary": false, "motionless_count": 0, "position_changes": 0, "current_zones": ["proche"], "entered_zones": ["proche"], "has_clip": false, "has_snapshot": false, "attributes": {}, "current_attributes": []}, "after": {"id": "1729980086.225972-v2uw7d", "camera": "AAA_000023", "frame_time": 1729980086.225972, "snapshot": {"frame_time": 1729980087.585224, "box": [811, 503, 839, 516], "area": 364, "region": [678, 400, 998, 720], "score": 0.7367673516273499, "attributes": []}, "label": "bird", "sub_label": null, "top_score": 0.7367673516273499, "false_positive": false, "start_time": 1729980086.225972, "end_time": 1729980095.874126, "score": 0.7367673516273499, "box": [811, 503, 839, 516], "area": 364, "ratio": 2.1538461538461537, "region": [678, 400, 998, 720], "stationary": false, "motionless_count": 0, "position_changes": 0, "current_zones": ["proche"], "entered_zones": ["proche"], "has_clip": false, "has_snapshot": false, "attributes": {}, "current_attributes": []}, "type": "end"}
```

Frigate config:
```
[...]

objects:
  track:
    - bird
    [...]

record:
  enabled: true
  retain:
    days: 30
    mode: all
  events:
    pre_capture: 15
    post_capture: 15
    retain:
      default: 30
      mode: all

snapshots:
  enabled: true
  crop: true
  bounding_box: true
  retain:
    default: 30

cameras:
  AAA_000023:
    ffmpeg:
      inputs:
        - input_args: preset-rtsp-restream-low-latency
          path: 
            rtsp://{FRIGATE_RTSP_USERNAME}:{FRIGATE_RTSP_PASSWORD}@HOST:8554/AAA_000023
          roles:
            - record
            - detect
    mqtt:
      required_zones:
        - proche
    zones:
      proche:
        coordinates: 0,0.111,1,0.362,1,1,0,1
        inertia: 1
        loitering_time: 0
    objects:
      track:
        - bird
```

**Comment by NickM-27** [comment]
> and MQTT event has has_clip: false and has_snapshot: false.

it means that the object does not have a snapshot or clip yet, so naturally there will be a 404

**Reply by fl42** [reply]
I understand that MQTT and the API are expected to be consistent.
However, what could explain the absence of a snapshot or clip from the event, even though recording and snapshots are enabled in the configuration?
I would expect all events to have `has_clip` and `has_snapshot` set to `true` (is [this](https://github.com/blakeblackshear/frigate/issues/3873#issuecomment-1252176801) still valid?)
Additionally, even after some time has passed since the event concluded, the clip or snapshot is still unavailable.

In fact, the API returns "Event not found," suggesting that the event may have disappeared?

To summarize:
- I am processing MQTT messages on the `events` topic. When `event["type"] == "new" and event["after"]["false_positive"]` is false, the event ID is recorded.
-  After the event has ended, I attempt to retrieve the snapshot or clip using the HTTP API, but it returns "Event not found."

What could explain this behavior?

Thank you!

**Reply by NickM-27** [reply]
There are many filters before an object will have a snapshot / clip saved. In this case it has none because it has not moved (position_changes 0)

If it has_snapshot then the API should work. We'd need to see an mqtt message and the API request / response if that is not the case.

**Reply by fl42** [reply]
Thank you for the pointer!

You're correct that a snapshot is not saved if positions_changes equals 0 (as noted [here](https://github.com/blakeblackshear/frigate/blob/v0.14.1/frigate/object_processing.py#L1046)). Does this imply that if an object appears but remains stationary, no snapshot or clip will be saved?

Additionally, the original issue may lie elsewhere. I receive an "Event not found" error when querying the API, rather than a "Snapshot not available" message (referenced [here](https://github.com/blakeblackshear/frigate/blob/v0.14.1/frigate/api/media.py#L942)). What could explain why an event published to MQTT does not appear in the API or SQLite database?

The MQTT messages:
```
{"before": {"id": "1730002973.093933-yzxw2i", "camera": "AAA_000009", "frame_time": 1730002973.093933, "snapshot": null, "label": "boat", "sub_label": null, "top_score": 0.0, "false_positive": true, "start_time": 1730002973.093933, "end_time": null, "score": 0.6814835667610168, "box": [1233, 305, 1251, 339], "area": 612, "ratio": 0.5294117647058824, "region": [960, 162, 1280, 482], "stationary": false, "motionless_count": 0, "position_changes": 0, "current_zones": [], "entered_zones": [], "has_clip": false, "has_snapshot": false, "attributes": {}, "current_attributes": []}, "after": {"id": "1730002973.093933-yzxw2i", "camera": "AAA_000009", "frame_time": 1730002973.39336, "snapshot": {"frame_time": 1730002973.39336, "box": [1233, 322, 1251, 339], "area": 306, "region": [960, 322, 1280, 642], "score": 0.6031749844551086, "attributes": []}, "label": "boat", "sub_label": null, "top_score": 0.7224900126457214, "false_positive": false, "start_time": 1730002973.093933, "end_time": null, "score": 0.6031749844551086, "box": [1233, 322, 1251, 339], "area": 306, "ratio": 1.0588235294117647, "region": [960, 322, 1280, 642], "stationary": false, "motionless_count": 1, "position_changes": 0, "current_zones": [], "entered_zones": [], "has_clip": false, "has_snapshot": false, "attributes": {}, "current_attributes": []}, "type": "new"}
{"before": {"id": "1730002973.093933-yzxw2i", "camera": "AAA_000009", "frame_time": 1730002973.39336, "snapshot": {"frame_time": 1730002973.39336, "box": [1233, 322, 1251, 339], "area": 306, "region": [960, 322, 1280, 642], "score": 0.6031749844551086, "attributes": []}, "label": "boat", "sub_label": null, "top_score": 0.7224900126457214, "false_positive": false, "start_time": 1730002973.093933, "end_time": null, "score": 0.6031749844551086, "box": [1233, 322, 1251, 339], "area": 306, "ratio": 1.0588235294117647, "region": [960, 322, 1280, 642], "stationary": false, "motionless_count": 1, "position_changes": 0, "current_zones": [], "entered_zones": [], "has_clip": false, "has_snapshot": false, "attributes": {}, "current_attributes": []}, "after": {"id": "1730002973.093933-yzxw2i", "camera": "AAA_000009", "frame_time": 1730002973.39336, "snapshot": {"frame_time": 1730002973.39336, "box": [1233, 322, 1251, 339], "area": 306, "region": [960, 322, 1280, 642], "score": 0.6031749844551086, "attributes": []}, "label": "boat", "sub_label": null, "top_score": 0.7224900126457214, "false_positive": false, "start_time": 1730002973.093933, "end_time": 1730002974.824336, "score": 0.6031749844551086, "box": [1233, 322, 1251, 339], "area": 306, "ratio": 1.0588235294117647, "region": [960, 322, 1280, 642], "stationary": false, "motionless_count": 1, "position_changes": 0, "current_zones": [], "entered_zones": [], "has_clip": false, "has_snapshot": false, "attributes": {}, "current_attributes": []}, "type": "end"}
```

The API response for `GET /api/events/1730002973.093933-yzxw2i/snapshot.jpg`:
`{"message":"Event not found","success":false}`

Surprisingly, I received "Event not found" despite the fact the event was fired in MQTT?
I expect to get "Snapshot not available".

Also, is it possible to ensure at least one snapshot/clip is saved even if the object does not move?

Thanks again for your support.

**Reply by NickM-27** [reply]
> What could explain why an event published to MQTT does not appear in the API or SQLite database?

every object that is detected is sent to MQTT regardless of if an event is saved in the DB or not

> Also, is it possible to ensure at least one snapshot/clip is saved even if the object does not move?

no, this would just result in lots of false positives

**Reply by fl42** [reply]
> every object that is detected is sent to MQTT regardless of if an event is saved in the DB or not

This is likely where I'm going wrong!
How can I filter MQTT messages to process only those related to "real" events (i.e., events saved to DB)?
If possible, I would prefer to stay at object level rather than relying on the `reviews` topic (because of the notification workflow I have).

**Reply by NickM-27** [reply]
use the has_snapshot and has_clip fields

**Reply by fl42** [reply]
That makes perfect sense!
I’ll adjust my workflow and let you know in a few days if everything is working smoothly.
Thank you for your help!
